cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -1173,5 +1175,173 @@ TEST(Substrait, JoinPlanInvalidKeys) {
+
+  ASSERT_OK_AND_ASSIGN(std::string dir_string,
+                       arrow::internal::GetEnvVar("PARQUET_TEST_DATA"));
+  auto file_name =
+      arrow::internal::PlatformFilename::FromString(dir_string)->Join("binary.parquet");
+
+  std::vector<fs::FileInfo> files;
+  const std::vector<std::string> f_paths = {file_name->ToString()};
+
+  for (const auto& f_path : f_paths) {
+    ASSERT_OK_AND_ASSIGN(auto f_file, filesystem->GetFileInfo(f_path));
+    files.push_back(std::move(f_file));
+  }
+
+  ASSERT_OK_AND_ASSIGN(auto ds_factory, dataset::FileSystemDatasetFactory::Make(
+                                            std::move(filesystem), std::move(files),
+                                            std::move(format), {}));
+  ASSERT_OK_AND_ASSIGN(auto dataset, ds_factory->Finish(dummy_schema));
+
+  auto options = std::make_shared<dataset::ScanOptions>();
+  options->projection = compute::project({}, {});
+  auto scan_node_options = dataset::ScanNodeOptions{dataset, options};
+
+  auto scan_declaration = compute::Declaration({"scan", scan_node_options});
+
+  ASSERT_OK_AND_ASSIGN(auto serialized_rel,
+                       SerializeRelation(scan_declaration, &ext_set));
+  ASSERT_OK_AND_ASSIGN(auto deserialized_decl,
+                       DeserializeRelation(*serialized_rel, ext_set));
+
+  auto dataset_comparator = [](std::shared_ptr<dataset::Dataset> ds_lhs,
+                               std::shared_ptr<dataset::Dataset> ds_rhs) -> bool {
+    const auto& fds_lhs = checked_cast<const dataset::FileSystemDataset&>(*ds_lhs);
+    const auto& fds_rhs = checked_cast<const dataset::FileSystemDataset&>(*ds_lhs);
+    const auto& files_lhs = fds_lhs.files();
+    const auto& files_rhs = fds_rhs.files();
+
+    bool cmp_fsize = files_lhs.size() == files_rhs.size();
+    uint64_t fidx = 0;
+    for (const auto& l_file : files_lhs) {
+      if (l_file != files_rhs[fidx]) {
+        return false;
+      }
+      fidx++;
+    }
+    bool cmp_file_format = fds_lhs.format()->Equals(*fds_lhs.format());
+    bool cmp_file_system = fds_lhs.filesystem()->Equals(fds_rhs.filesystem());
+    return cmp_fsize && cmp_file_format && cmp_file_system;
+  };
+
+  auto scan_option_comparator = [dataset_comparator](
+                                    const dataset::ScanNodeOptions& lhs,
+                                    const dataset::ScanNodeOptions& rhs) -> bool {
+    bool cmp_rso = lhs.require_sequenced_output == rhs.require_sequenced_output;
+    bool cmp_ds = dataset_comparator(lhs.dataset, rhs.dataset);
+    return cmp_rso && cmp_ds;
+  };
+
+  EXPECT_EQ(deserialized_decl.factory_name, scan_declaration.factory_name);
+  const auto& lhs =
+      checked_cast<const dataset::ScanNodeOptions&>(*deserialized_decl.options);
+  const auto& rhs =
+      checked_cast<const dataset::ScanNodeOptions&>(*scan_declaration.options);
+  ASSERT_TRUE(scan_option_comparator(lhs, rhs));
+#endif
+}
+
+TEST(Substrait, SerializeRelationEndToEnd) {
```

Comment:
What is this test adding that the previous test does not already cover?

Reply:
Yes indeed, I was thinking the same about these comparators. I cc`ed you somewhere else regarding this. I can create a JIRA to add the equals methods separately and I can work on them. 

And if we used it again definitely we should push this for a util. I will add a note there.

Reply:
The previous test just compares the configs, this test is an end-to-end test which considers the output data from a given exec plan and ser/de exec plans output. It is just a validation on keeping the functionality intact.